src/Indigo/Contracts/Holdings/Types.hs

```diff
@@ -0,0 +1,98 @@
+-- SPDX-FileCopyrightText: 2020 TBD
+--
+-- SPDX-License-Identifier: LicenseRef-Proprietary
+module Indigo.Contracts.Holdings.Types
+  ( Parameter (..)
+  , Storage
+  , StorageFields (..)
+  , TokenMeta (..)
+  , dummyMeta
+  , mkStorage
+  )
+where
+
+import Indigo
+
+import Fmt (Buildable(..), (+|), (|+))
+
+import qualified Lorentz.Contracts.ManagedLedger.Types as ML
+import qualified Lorentz.Contracts.Spec.ManagedLedgerInterface as ML
+import Lorentz.TypeAnns (HasTypeAnn)
+
+data StorageFields = StorageFields
+  { tokenMeta :: TokenMeta
+  , mbSafelistAddress :: Maybe Address
+  , owner :: Address
+  , admin :: Address
+  , mbNewAdmin :: Maybe Address
+  , paused :: Bool
+  , transferable :: Bool
+  , totalSupply :: Natural
+  }
+  deriving stock Generic
+  deriving anyclass IsoValue
+
+data TokenMeta = TokenMeta
+  { tmName :: MText
+  , tmSymbol :: MText
+  , tmId :: MText
+  }
+  deriving stock (Eq, Generic)
+  deriving anyclass (IsoValue, HasTypeAnn)
+
+instance Buildable TokenMeta where
+  build TokenMeta{..} =
+    "Token meta: name = '" +| tmName |+ "', symbol = '" +| tmSymbol |+
+    "', id = '" +| tmId |+ "'"
+
+instance TypeHasDoc TokenMeta where
+  typeDocMdDescription = "Meta information about token."
+
+dummyMeta :: TokenMeta
+dummyMeta = TokenMeta [mt|KekToken|] [mt|Kek|] [mt|Ququareq|]
+
+instance HasFieldOfType StorageFields name field =>
+         StoreHasField StorageFields name field where
+  storeFieldOps = storeFieldOpsADT
+
+type Storage = ML.StorageSkeleton StorageFields
+
+mkStorage :: Address -> Maybe Address -> TokenMeta -> Storage
+mkStorage owner mbSafelist meta =
+  ML.mkStorageSkeleton mempty fields
+  where
+    fields :: StorageFields
+    fields = StorageFields
+      { tokenMeta = meta
+      , mbSafelistAddress = mbSafelist
+      , owner = owner
+      , admin = owner
+      , mbNewAdmin = Nothing
+      , paused = False
+      , transferable = True
+      , totalSupply = 0
+      }
+
+data Parameter
+  = SetName MText
```

Comment:
Does not match specification (not sure if intentional) as in there this has `:newName`, same for `SetSymbol`

Reply:
Unfortunately, you cannot easily add named label to single value in the parameter type :(
In this case `instance ParameterHasEntryPoints Parameter` complains.

Also, I believe it's fine to drop `:newName` and update the specification.

Reply:
```haskell
  = SetName ("newName" :! MText)
```
should work fine, I don't think `ParameterHasEntryPoints`'s instance would have problems with that.
It also should not create any other problems, you just need to use `UnName #newName newName` when you have to use that parameter.

If changing the specification is fine as well, you can do that instead if you prefer so. Your call.

Reply:
```
    • Couldn't match type ‘MText’
                     with ‘named-0.3.0.1:Named.Internal.NamedF Identity MText "newName"’
        arising from a functional dependency between:
          constraint ‘StoreHasField
                        StorageFields
                        "tokenName"
                        (named-0.3.0.1:Named.Internal.NamedF Identity MText "newName")’
            arising from a use of ‘setStorageField’
          instance ‘StoreHasField StorageFields "tokenName" MText’
            at src/Indigo/Contracts/Holdings/Types.hs:52:10-54
    • In a stmt of a 'do' block:
        setStorageField @Storage #tokenName newName
      In the expression:
        do doc $ DDescription "Change token name."
           ensureSenderIsAdmin
           ensureNotPaused
           setStorageField @Storage #tokenName newName
      In the second argument of ‘(//->)’, namely
        ‘\ newName
           -> do doc $ DDescription "Change token name."
                 ensureSenderIsAdmin
                 ....’
   |        
55 |         setStorageField @Storage #tokenName newName
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Oh, you're right